src/rgw/store/dbstore/dbstore_mgr.h

```diff
@@ -12,26 +12,50 @@
 #include "common/ceph_context.h"
 #include "common/dbstore.h"
 #include "sqlite/sqliteDB.h"
+#include <boost/lockfree/queue.hpp>
 
 using namespace std;
 using namespace rgw::store;
 using DB = rgw::store::DB;
 
 /* XXX: Should be a dbstore config option */
 const static string default_tenant = "default_ns";
+#define MAX_QUEUE_DEFAULT 20 // make it configurable
 
 using namespace std;
 
+typedef boost::lockfree::queue<DB*, boost::lockfree::fixed_sized<true>> DBStoreQueue;
+
 class DBStoreManager {
 private:
   map<string, DB*> DBStoreHandles;
   DB *default_db = NULL;
   CephContext *cct;
 
+  // used in the dtor for dbstore conn cleanup
+  static std::atomic<uint64_t> max_conn; // XXX: make it configurable
+  static std::atomic<uint64_t> total_conn; // total connections created so far
+
+  static DBStoreQueue db_connections;
+  static std::mutex db_mutex;
+  static std::condition_variable db_cond;
```

Comment:
> in general, it's not quite clear what the `DB` abstraction should represent. in terms of `map<string, DB*> DBStoreHandles`, it was the per-tenant database file and its corresponding connection. in this PR it changes to mean the database connection specifically
> 

Right. I was trying to draft these changes for default tenant first and then extend to per-tenant pool. But as you mentioned, it breaks the current assumption of having per-tenant db file. Will address it.

> are you sure this is the right split? don't we still want some state that's per-database file? as mentioned, `static` isn't a good fit for `db_connections` because they can't be shared between files. so maybe `DB` should remain as this per-database abstraction, and let `SQLiteDB` manage its connection pool internally?

Could you please share more insights on how it would benefit by making it internal to SQLiteDB. Even then we need to create unique connection pool for each tenant right (but in SQLiteDB layer)?
I haven't checked in depth how other DB backends (esp., Postgres) would behave with concurrent connections. But having the pool management in common layer might benefit all ? Or maybe just define the routines in common files which can be later be used by any backend?

Reply:
> Could you please share more insights on how it would benefit by making it internal to SQLiteDB. Even then we need to create unique connection pool for each tenant right (but in SQLiteDB layer)?

the complexity needs to live somewhere, it's just a question of what layers need access to it. if the upper layer (`DBStore` in this case) doesn't need to interact with connection pooling, then it's better to hide it under the `DB` abstraction

but if `DBStore` does care about pooling - for example, if it needs the ability to execute several `DB` calls over the same connection to implement some compound operation - then we definitely do need to expose it like you have here

> I haven't checked in depth how other DB backends (esp., Postgres) would behave with concurrent connections. But having the pool management in common layer might benefit all ? Or maybe just define the routines in common files which can be later be used by any backend?

right, i agree that connection pooling is likely general to most/all backends, and we should share the common code between them. i'm just suggesting that we don't bake it into the `DB` interface unless it's a feature that `DBStore` needs to use it correctly

Reply:
got it. Thanks.. will look into that aspect.

Reply:
I'm not entirely sure about this.  one thing I am sure of is that sqlite3 scales very differently with threads/sessions than most other systems, and I'm not convinced designing around sqlite3 is really apt;  I need to focus on sessions in scaling the postgresql implementation, so maybe we could wait until we get a bit more information from that?